crates/uv-platform-tags/src/abi_tag.rs

```diff
@@ -0,0 +1,445 @@
+    /// Ex) `none`
+    None,
+    /// Ex) `abi3`
+    Abi3,
+    /// Ex) `cp39m`, `cp310t`
+    CPython {
+        gil_disabled: bool,
+        python_version: (u8, u8),
+    },
+    /// Ex) `pypy39_pp73`
+    PyPy {
+        python_version: (u8, u8),
+        implementation_version: (u8, u8),
+    },
+    /// Ex) `graalpy310_graalpy240_310_native`
+    GraalPy {
+        python_version: (u8, u8),
+        implementation_version: (u8, u8),
+    },
+    /// Ex) `pyston38-pyston_23`
+    Pyston {
+        python_version: (u8, u8),
+        implementation_version: (u8, u8),
+    },
+}
+
+impl std::fmt::Display for AbiTag {
+    /// Format an [`AbiTag`] as a string.
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::None => write!(f, "none"),
+            Self::Abi3 => write!(f, "abi3"),
+            Self::CPython {
+                gil_disabled,
+                python_version: (major, minor),
+            } => {
+                if *minor <= 7 {
+                    write!(f, "cp{major}{minor}m")
+                } else if *gil_disabled {
+                    // https://peps.python.org/pep-0703/#build-configuration-changes
+                    // Python 3.13+ only, but it makes more sense to just rely on the sysconfig var.
+                    write!(f, "cp{major}{minor}t")
+                } else {
+                    write!(f, "cp{major}{minor}")
+                }
+            }
+            Self::PyPy {
+                python_version: (py_major, py_minor),
+                implementation_version: (impl_major, impl_minor),
+            } => {
+                write!(f, "pypy{py_major}{py_minor}_pp{impl_major}{impl_minor}")
+            }
+            Self::GraalPy {
+                python_version: (py_major, py_minor),
+                implementation_version: (impl_major, impl_minor),
+            } => {
+                write!(
+                    f,
+                    "graalpy{py_major}{py_minor}_graalpy{impl_major}{impl_minor}_{py_major}{py_minor}native"
+                )
+            }
+            Self::Pyston {
+                python_version: (py_major, py_minor),
+                implementation_version: (impl_major, impl_minor),
+            } => {
+                write!(
+                    f,
+                    "pyston{py_major}{py_minor}-pyston_{impl_major}{impl_minor}"
+                )
+            }
+        }
+    }
+}
+
+impl FromStr for AbiTag {
+    type Err = ParseAbiTagError;
+
+    /// Parse an [`AbiTag`] from a string.
+    #[allow(clippy::cast_possible_truncation)]
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        /// Parse a Python version from a string (e.g., convert `39` into `(3, 9)`).
+        fn parse_python_version(
+            version_str: &str,
+            implementation: &'static str,
+            full_tag: &str,
+        ) -> Result<(u8, u8), ParseAbiTagError> {
+            let major = version_str
+                .chars()
```

Comment:
I think below (`.get(1..)`) you are treating this as an ASCII string, so you might as well fully commit and elide UTF-8 decoding here.